src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,324 @@
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ * \file array/cpu/labor_pick.h
+ * \brief Template implementation for layerwise pick operators.
+ */
+
+#ifndef DGL_ARRAY_CPU_LABOR_PICK_H_
+#define DGL_ARRAY_CPU_LABOR_PICK_H_
+
+#include <dgl/random.h>
+#include <dgl/array.h>
+#include <dmlc/omp.h>
+#include <dgl/runtime/parallel_for.h>
+#include <parallel_hashmap/phmap.h>
+#include <functional>
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <numeric>
+#include <memory>
+#include <utility>
+#include <cmath>
+#include <iostream>
+#include <pcg_random.hpp>
+
+namespace dgl {
+namespace aten {
+namespace impl {
+
+constexpr double eps = 0.0001;
+
+template <typename IdxType, typename FloatType>
+void compute_importance_sampling_probabilities(
+  DGLContext ctx,
+  DGLDataType fidtype,
+  const IdxType max_degree,
+  const IdxType num_rows,
+  const int importance_sampling,
+  const bool weights,
+  const IdxType* rows_data,
+  const IdxType* indptr,
+  const FloatType* A,
+  const IdxType* indices,
+  const IdxType num_picks,
+  const FloatType* ds,
+  FloatType* cs,
+  phmap::flat_hash_map<IdxType, FloatType> &hop_map
+) {
+  constexpr FloatType ONE = 1;
+  // ps stands for \pi in arXiv:2210.13339
+  FloatArray ps_array = NDArray::Empty({max_degree + 1}, fidtype, ctx);
+  FloatType* ps = ps_array.Ptr<FloatType>();
+
+  double prev_ex_nodes = max_degree * num_rows;
+
+  phmap::flat_hash_map<IdxType, FloatType> hop_map2;
+  for (int iters = 0; iters < importance_sampling || importance_sampling < 0; iters++) {
+    /*
+    An implementation detail about when there are no weights, then the first c values can be computed in O(1) because all the edge weights are uniform. The c value is just k / d where k is fanout and d is the degree. But when there are weights, since the computed c values are incorrect, I have to skip some logic making use of it to compute the right c values below.
+
+    The later iterations will have correct c values so they will take the if in that case.
+
+    I also increase importance_sampling by 1 above so that the weighted case can do one more of this iteration to match the unweighted case.
+    */
+    if (!weights || iters) {
+      hop_map2.clear();
+      for (int64_t i = 0; i < num_rows; ++i) {
+        const FloatType c = cs[i];
+        const IdxType rid = rows_data[i];
+        for (auto j = indptr[rid]; j < indptr[rid + 1]; j++) {
+          const auto ct = c * (weights && iters == 1 ? A[j] : 1);
+          auto itb = hop_map2.emplace(indices[j], ct);
+          if (!itb.second)
+            itb.first->second = std::max(ct, itb.first->second);
+        }
+      }
+      if (hop_map.empty())
+        hop_map = std::move(hop_map2);
+      else
+        for (auto it : hop_map2)
+          hop_map[it.first] *= it.second;
+    }
```

Comment:
I see.  So this is to compute the `hop_map` (in particular the necessary $\pi_s$, in other words Eq 18).  I will just say
```
Update the pi array according to Eq 18.
```
By the way, two questions:

Should Eq.18 in the paper be written as $\pi_t^{(i+1)} = \pi_t^{(i)} \max_{t \to s} c_s^{(i+1)}$?

Could we structure the code so that it looks like what you have said in the comment, i.e. if I understood correctly, something like
```C++
for (int iters = 0; iters < importance_sampling || importance_sampling < 0; ++iters) {
  for (int64_t i = 0; i < num_rows; ++i) {
    if (iters == 0 && !weights) {
      cs[i] = num_picks / d;
    } else {
      // Update cs[i] as usual...
    }
  }
  // Check convergence...
}
```
This way you probably don't need to add one to ``importance_sampling`` as well?

Reply:
But the last thing one needs to do is to recompute c_s with the updated pi values. If importance_sampling is 0, then I can avoid calling this function alltogether along with whatever it requires in terms of resources. So, I would either have to break out of the loop after updating c_s with your suggestion, and it would still require some logic. I think it is good as it is for now.

Reply:
The equation in the paper is correct.

Reply:
> The equation in the paper is correct.

Yeah I realized that $c_s(\pi^{(i)})$ refers to the solution to Eq. (15) given $\pi$ so you are right.  Never mind.

> But the last thing one needs to do is to recompute c_s with the updated pi values. If importance_sampling is 0, then I can avoid calling this function alltogether along with whatever it requires in terms of resources. So, I would either have to break out of the loop after updating c_s with your suggestion, and it would still require some logic. I think it is good as it is for now.

OK.  I'll try to rephrase the description in the code comment above then.